Public/Src/Engine/Dll/Distribution/DistributionHelpers.cs

```diff
@@ -148,6 +150,35 @@ internal static string GetServiceName(string ipAddress, int port)
             return string.Format(CultureInfo.InvariantCulture, "{0}::{1}", ipAddress, port);
         }
 
+        internal static string GetExecuteDescription(IList<long> semiStableHashes)

```

Comment:
There is a lot of string concatenation and conversion going on. 
If this gets called a lot It would be more effecient to create a string builder and use StringBuilder.Format to the the x16 conversion so we don't allocate a string for each hash, and then another for them joined and then another that puts execute in front. 

Same applies for other method where even more string concatenation happens.

Again if this is for failed pips or once per build its okay, but if this is for every pip it is probably worth optimizing

Reply:
This is probably negligible when compared to the amount of string concatenation we do in the cache code and logging in general.

---
In reply to: [272719241](https://github.com/Microsoft/BuildXL/pull/91#discussion_r272719241) [](ancestors = 272719241)